python/ray/data/_internal/plan.py

```diff
@@ -329,6 +329,8 @@ def _optimize(self) -> Tuple[BlockList, DatasetStats, List[Stage]]:
         """
         context = DatasetContext.get_current()
         blocks, stats, stages = self._get_source_blocks_and_stages()
+        if context.optimize_reorder_stages:
+            stages = _reorder_stages(stages)
```

Comment:
@ericl If you rewrite the read stage before you reorder stages, won't the read->map_batches fusion work without the `lazy` vs `defer_execution` hack?

Reply:
No, because if you call .randomize_blocks() eagerly it will materialize the read blocks, so it's too late. We have to force it to be lazy as a special case.

I also need this for auto_repartition() anyway, so I think this is a sensible thing... and if we move to lazy by default it will go away.